lib/java/com/google/android/material/lists/item/Visual.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2020 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.android.material.lists.item;
+
+import android.content.Context;
+import android.util.AttributeSet;
+import android.view.Gravity;
+import android.widget.LinearLayout;
+
+import androidx.annotation.DimenRes;
+import androidx.annotation.Dimension;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import com.google.android.material.R;
+import com.google.android.material.imageview.ShapeableImageView;
+import com.google.android.material.internal.ViewUtils;
+import com.google.android.material.shape.CornerFamily;
+import com.google.android.material.shape.ShapeAppearanceModel;
+
+public class Visual extends ShapeableImageView implements TotalLinesListener {
+
+  protected Visual(Context context) {
+    super(context);
+  }
+
+  public Visual(Context context, @Nullable AttributeSet attrs) {
+    super(context, attrs);
+  }
+
+  public Visual(Context context, @Nullable AttributeSet attrs, int defStyle) {
+    super(context, attrs, defStyle);
+  }
+
+  public static final int NONE = 0;
+  public static final int ICON = 1;
+  public static final int CIRCLE = 2;
+  public static final int SQUARE = 3;
+  public static final int RECTANGLE = 4;
+
+  private int type = NONE;
+
+  @NonNull
+  public int getType() {
+    return type;
+  }
+
+  public void setType(int type) {
+
+    this.type = type;
+    onTotalLinesChange(DEFAULT);
```

Comment:
This seems strange to me - using `onTotalLinesChange` as the single method where layout updates occur doesn't seem right. Setting the image type doesn't affect the total lines. I think a separate method is needed for overall layout updates is needed, or rather split up `onTotalLinesChange` accordingly.

The same goes for the other classes that do this.

Reply:
I believe only Visual implemented this awkward construction, however I can confirm that it is now removed from all files. Thanks for the update, the code is definitely easier to read.